### PR TITLE
v2 Hook rework

### PR DESCRIPTION
**Pull request by IljaDaderko**
This one took a while as I iterated on this for a while 😅 but managed to get almost all wagmi hooks (useContractReads and useContractInfiniteReads) done with few util hooks + all logic is routed through our controllers.

## Changes

1. Most of the hooks no return `data, error, isLoading, refetch` components (sometimes named differently where it makes sense), that should allow usage within any wagmi project (I've not included other returns like `isFetching` / `isSuccess` as these can be deducted from above, but if there is demand for them we can add them in no problem).

2. All hooks are now chain agnostic as well i.e. if you don't provide chainId it will use currently connected one

3. Hooks that are `watchable` and `enablable` in wagmi are now also the same in our react package. Where watcher util was not available from wagmi a fallback `watchBlockNumber` is used

4. All types are now it one place and deducted from ethereum api file

5. All hooks have example sections

6. Yarn v1 is now used as package manager (just results in less issues within monorepo/workspaces compared to npm)

7. React package no longer depends on ethereum one, as everything is routed through core (sets us up for easier solana expansion in the future)

8. Added few custom chains as poc for future customisation (avalanche, fantom and binance alongside their testnets)

9. If user switches account / network in the wallet, all hooks react to the change (unless they have set explicit chainId)

10. We no longer use "wrapper provider" in react, rather just component that can be thrown anywhere, usually at the bottom of the app. Should help save on few re-renders / decouple modal stuff.

## Todo

Need to update readme, but that will come in next pr, otherwise this one will be even harder to review

### CODE REVIEW

**Review by devceline — APPROVED** [review]
First of all, phenomenal work. Kudos.

Just left some questions. 
- Why are we removing our own functions?
- What's the motivation with putting all the types in one file?
- Is there a reason we're going with hardcodded ens values in the example?

File: chains/ethereum/src/utils/wagmiTools.ts [hunk 1]
```diff
@@ -55,6 +74,13 @@ export function defaultConnectors({ appName, chains: connectorChains }: GetDefau
       chains: connectorChains,
       options: { appName, headlessMode: true }
     }),
-    new MetaMaskConnector({ chains: connectorChains })
+    new MetaMaskConnector({
+      chains: connectorChains,
+      options: {
+        shimDisconnect: true,
+        shimChainChangedDisconnect: false,
+        UNSTABLE_shimOnConnectSelectAccount: true
```
[the remark this answers is not in the document]
Comment (IljaDaderko) [reply chains/ethereum/src/utils/wagmiTools.ts:82]: Same config as in connectkit and rainbowkit, and yeh these are mentioned in wagmi doc.

File: chains/ethereum/package.json [hunk 2]
```diff
@@ -17,12 +17,14 @@
     "typecheck": "tsc --noEmit"
   },
   "dependencies": {
-    "@coinbase/wallet-sdk": "3.5.2",
-    "@wagmi/core": "0.5.6",
+    "@coinbase/wallet-sdk": "3.5.3",
+    "@wagmi/core": "0.0.0-20220928064342",
```
[the remark this answers is not in the document]
Comment (IljaDaderko) [reply chains/ethereum/package.json:21]: This one has fix for `watchBlockNumber` function, where it's unsubsribe return didn't work. It is not up on normal release track atm.

File: packages/core/src/controllers/statefull/ClientCtrl.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  ClientCtrlSetEthereumClientArgs,
+  ClientCtrlState
+} from '../../../types/statefullCtrlTypes'
+
+// -- initial state ------------------------------------------------ //
+const state = proxy<ClientCtrlState>({
+  initialized: false,
+  ethereum: undefined
+})
+
```
[the remark this answers is not in the document]
Comment (IljaDaderko) [reply packages/core/src/controllers/statefull/ClientCtrl.ts:12]: Yep, and solana will have it's own config similar to ethereum. In theory both can exist together even, but I don't think there are apps that do that.

File: packages/core/src/controllers/stateless/AccountCtrl.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import type { AccountCtrlWatchCallback } from '../../../types/statelessCtrlTypes'
+import { ClientCtrl } from '../statefull/ClientCtrl'
+
+export const AccountCtrl = {
+  watch(_options: undefined, callback: AccountCtrlWatchCallback) {
+    const unwatch = ClientCtrl.ethereum().watchAccount(callback)
```
[the remark this answers is not in the document]
Comment (IljaDaderko) [reply packages/core/src/controllers/stateless/AccountCtrl.ts:6]: Honestly, in this pr, I just attempted to get clients decoupled from react package into controllers. Next step would be figuring out what you mentioned, but for that we will prob need to see how solib shapes up / how different they are etc.

File: chains/ethereum/package.json (hunk 2 above)
Comment (crypblizz8) [inline chains/ethereum/package.json:21]: Not going to break?

Comment (IljaDaderko) [reply chains/ethereum/package.json:21]: Should be good. It's based on 0.5.7 release and implements a fix that we discussed with wagmi team, which will be part of next release.

File: examples/react/src/pages/_app.tsx [hunk 5]
```diff
@@ -16,15 +16,23 @@ const modalConfig: ConfigOptions = {
...
   }
 }
 
 export default function App({ Component, pageProps }: AppProps) {
   return (
-    <Web3ModalProvider config={modalConfig}>
+    <>
       <Component {...pageProps} />
-    </Web3ModalProvider>
+      <Web3Modal config={modalConfig} />
```
Comment (crypblizz8) [inline examples/react/src/pages/_app.tsx:35]: This is starting to look good, great job!

**Comment by crypblizz8** [comment]
Amazing feat!

- Really like how things are more abstracted to the top with Cntrls. 
- Stateless and Stateful separation with Types makes sense. 
- Provider removal makes easier. 
- Yarn Switch
- LGTM to merge.. Ready to try and integrate to other apps and see how it works in prod.

**Pull request merged**